homeassistant/components/knx/__init__.py

```diff
@@ -170,36 +172,55 @@
 )
 
 
-async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
-    """Set up the KNX integration."""
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Load a config entry."""
+    conf = hass.data.get(DATA_KNX_CONFIG)
+
+    # If user didn't have configuration.yaml config, generate defaults
+    if conf is None:
+        conf = CONFIG_SCHEMA({DOMAIN: dict(entry.data)})[DOMAIN]
+    elif any(key in conf for key in entry.data):
+        shared_keys = conf.keys() & entry.data.keys()
+        override = {k: entry.data[k] for k in shared_keys}
+        _LOGGER.info(
+            "Data in your configuration entry is going to override your "
+            "configuration.yaml: %s",
+            override,
+        )
+
+    config = {**entry.data, **conf}
+
     try:
         knx_module = KNXModule(hass, config)
         hass.data[DOMAIN] = knx_module
         await knx_module.start()
     except XKNXException as ex:
-        _LOGGER.warning("Could not connect to KNX interface: %s", ex)
-        hass.components.persistent_notification.async_create(
-            f"Could not connect to KNX interface: <br><b>{ex}</b>", title="KNX"
-        )
+        raise ConfigEntryNotReady from ex
```

Comment:
Only store in `hass.data` after config entry will be set up.

```suggestion
        raise ConfigEntryNotReady from ex
    else:
        hass.data[DOMAIN] = knx_module
```

Reply:
btw, why not do `hass.data[DOMAIN][config_entry.entry_id]` to support 2 KNX gateways?

Reply:
I didn't see a reason to do this right now and decided against it since this PR is already gigantic. AFAIK we can still do that later on - right now it wouldn't work as we only allow a single instance and our config flows dont have a unique id.